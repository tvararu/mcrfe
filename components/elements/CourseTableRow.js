import Link from "next/link";
import CourseStatusTag from "../elements/CourseStatusTag";

const siteIsFindable = ({ publish, status }) =>
  publish === "Y" && status === "R";

const courseIsFindable = course =>
  course.courseSitesByCourseId.nodes.some(siteIsFindable);

const courseIsNew = course =>
  course.courseSitesByCourseId.nodes.some(({ status }) => status === "N");

const courseUcasStatus = course => {
  if (courseIsFindable(course)) return "Running";
  if (courseIsNew(course)) return "New â€“ not yet running";

  return "Not running";
};

const courseIsRunning = course => courseUcasStatus(course) === "Running";

const siteOpenForApplications = site =>
  site.applicationsAcceptedFrom &&
  new Date(site.applicationsAcceptedFrom) < new Date();

const siteHasVacancies = site => site.vacStatus !== "";

const courseOpenForApplications = course =>
  course.courseSitesByCourseId.nodes
    .filter(siteIsFindable)
    .filter(siteOpenForApplications)
    .some(siteHasVacancies);

const courseHasVacancies = course =>
  course.courseSitesByCourseId.nodes
    .filter(siteIsFindable)
    .some(siteHasVacancies);

export default ({ providerCode, course, latestEnrichment }) => (
  <tr className="govuk-table__row">
    <td className="govuk-table__cell">
      <a
        href="#"
        className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
      >
        {course.name} ({course.courseCode})
      </a>
      <span className="govuk-body-s">QTS full time</span>
    </td>
    <td className="govuk-table__cell">{courseUcasStatus(course)}</td>
    <td className="govuk-table__cell">
      {courseIsRunning(course) && (
        <CourseStatusTag enrichment={latestEnrichment} />
      )}
    </td>
    <td className="govuk-table__cell">
      {courseIsFindable(course) ? (
        <a
          className="govuk-link"
          href={`https://bat-dev-search-and-compare-ui-app.azurewebsites.net/course/${providerCode}/${
            course.courseCode
          }`}
        >
          Yes - view online
        </a>
      ) : (
        "No"
      )}
    </td>
    <td className="govuk-table__cell">
      {courseIsRunning(course)
        ? courseOpenForApplications(course)
          ? "Open"
          : "Closed"
        : ""}
    </td>
    <td className="govuk-table__cell">
      {courseIsRunning(course) ? (
        <>
          {courseHasVacancies(course) ? "Yes" : "No"} (
          <a
            className="govuk-link"
            href={`/organisations/${providerCode}/courses/${
              course.courseCode
            }/vacancies`}
          >
            Edit
          </a>
          )
        </>
      ) : (
        ""
      )}
    </td>
  </tr>
);
