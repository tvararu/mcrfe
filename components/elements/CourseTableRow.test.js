import renderer from "react-test-renderer";
import CourseTableRow from "./CourseTableRow";
jest.mock("next/link");

const courses = [
  {
    _label: "running and open",
    courseCode: "X100",
    providerByAccreditingProviderId: null,
    name: "Primary",
    courseSitesByCourseId: {
      nodes: [
        {
          publish: "Y",
          status: "R",
          applicationsAcceptedFrom: "2018-10-09",
          vacStatus: "F"
        }
      ]
    }
  },
  {
    _label: "running and closed",
    courseCode: "X100",
    providerByAccreditingProviderId: null,
    name: "Primary",
    courseSitesByCourseId: {
      nodes: [
        {
          publish: "Y",
          status: "R",
          applicationsAcceptedFrom: "2018-10-09",
          vacStatus: ""
        }
      ]
    }
  },
  {
    _label: "new",
    courseCode: "X100",
    providerByAccreditingProviderId: null,
    name: "Primary",
    courseSitesByCourseId: {
      nodes: [
        {
          publish: "Y",
          status: "N",
          applicationsAcceptedFrom: "2018-10-09",
          vacStatus: "F"
        }
      ]
    }
  },
  {
    _label: "not running",
    courseCode: "X100",
    providerByAccreditingProviderId: null,
    name: "Primary",
    courseSitesByCourseId: {
      nodes: [
        {
          publish: "N",
          status: "S",
          applicationsAcceptedFrom: "2018-10-09",
          vacStatus: "F"
        }
      ]
    }
  }
];

describe("CourseTableRow", () => {
  courses.forEach(course => {
    it(`renders correctly with ${course._label} course`, () => {
      const component = renderer.create(
        <CourseTableRow providerCode="A01" course={course} />
      );
      const tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
