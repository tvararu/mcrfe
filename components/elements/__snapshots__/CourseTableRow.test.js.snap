// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseTableRow renders correctly with new course 1`] = `
<tr
  className="govuk-table__row"
>
  <td
    className="govuk-table__cell"
  >
    <a
      className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
      href="#"
    >
      Primary
       (
      X100
      )
    </a>
    <span
      className="govuk-body-s"
    >
      QTS full time
    </span>
  </td>
  <td
    className="govuk-table__cell"
  >
    New â€“ not yet running
  </td>
  <td
    className="govuk-table__cell"
  />
  <td
    className="govuk-table__cell"
  >
    No
  </td>
  <td
    className="govuk-table__cell"
  >
    
  </td>
  <td
    className="govuk-table__cell"
  >
    
  </td>
</tr>
`;

exports[`CourseTableRow renders correctly with not running course 1`] = `
<tr
  className="govuk-table__row"
>
  <td
    className="govuk-table__cell"
  >
    <a
      className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
      href="#"
    >
      Primary
       (
      X100
      )
    </a>
    <span
      className="govuk-body-s"
    >
      QTS full time
    </span>
  </td>
  <td
    className="govuk-table__cell"
  >
    Not running
  </td>
  <td
    className="govuk-table__cell"
  />
  <td
    className="govuk-table__cell"
  >
    No
  </td>
  <td
    className="govuk-table__cell"
  >
    
  </td>
  <td
    className="govuk-table__cell"
  >
    
  </td>
</tr>
`;

exports[`CourseTableRow renders correctly with running and closed course 1`] = `
<tr
  className="govuk-table__row"
>
  <td
    className="govuk-table__cell"
  >
    <a
      className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
      href="#"
    >
      Primary
       (
      X100
      )
    </a>
    <span
      className="govuk-body-s"
    >
      QTS full time
    </span>
  </td>
  <td
    className="govuk-table__cell"
  >
    Running
  </td>
  <td
    className="govuk-table__cell"
  >
    <div
      className="jsx-4100856793 govuk-tag tag--small empty"
    >
      Empty
    </div>
  </td>
  <td
    className="govuk-table__cell"
  >
    <a
      className="govuk-link"
      href="https://bat-dev-search-and-compare-ui-app.azurewebsites.net/course/A01/X100"
    >
      Yes - view online
    </a>
  </td>
  <td
    className="govuk-table__cell"
  >
    Closed
  </td>
  <td
    className="govuk-table__cell"
  >
    No
     (
    <a
      className="govuk-link"
      href="/organisations/A01/courses/X100/vacancies"
    >
      Edit
    </a>
    )
  </td>
</tr>
`;

exports[`CourseTableRow renders correctly with running and open course 1`] = `
<tr
  className="govuk-table__row"
>
  <td
    className="govuk-table__cell"
  >
    <a
      className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
      href="#"
    >
      Primary
       (
      X100
      )
    </a>
    <span
      className="govuk-body-s"
    >
      QTS full time
    </span>
  </td>
  <td
    className="govuk-table__cell"
  >
    Running
  </td>
  <td
    className="govuk-table__cell"
  >
    <div
      className="jsx-4100856793 govuk-tag tag--small empty"
    >
      Empty
    </div>
  </td>
  <td
    className="govuk-table__cell"
  >
    <a
      className="govuk-link"
      href="https://bat-dev-search-and-compare-ui-app.azurewebsites.net/course/A01/X100"
    >
      Yes - view online
    </a>
  </td>
  <td
    className="govuk-table__cell"
  >
    Open
  </td>
  <td
    className="govuk-table__cell"
  >
    Yes
     (
    <a
      className="govuk-link"
      href="/organisations/A01/courses/X100/vacancies"
    >
      Edit
    </a>
    )
  </td>
</tr>
`;
